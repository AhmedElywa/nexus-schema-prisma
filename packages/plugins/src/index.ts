export * from './select';
export * from './sdlInputs';
