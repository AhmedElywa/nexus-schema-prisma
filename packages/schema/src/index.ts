export * from '@paljs/types';
export * from './json';
export * from './camelCase';
export * from './typescript';
export * from './PrismaReader';
