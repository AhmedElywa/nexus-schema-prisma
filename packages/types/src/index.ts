export * from './cli';
export * from './generatedSchema';
export * from './generator';
export * from './configFile';
export * from './adminSchema';
export * from './dmmf';
