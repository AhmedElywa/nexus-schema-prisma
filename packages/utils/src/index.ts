export * from './dmmf';
export * from './schemaInputs';
export { getConfig, getEnvPaths, tryLoadEnvs, getDMMF } from '@prisma/internals';
