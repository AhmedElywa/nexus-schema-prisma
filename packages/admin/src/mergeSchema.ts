import { Schema, SchemaModel, SchemaField } from './types';
import { SchemaObject, Model, Field } from '@prisma-tools/schema';

export function mergeSchema(object: SchemaObject, schema: Schema): Schema {
  const newSchema: Schema = {
    models: [],
    enums: object.enums,
  };
  object.models.forEach((item) => {
    const schemaItem = schema.models.find((model) => model.id === item.name);
    if (!schemaItem) {
      if (checkIdFieldExist(item)) {
        newSchema.models.push(handleNewModel(item));
      }
    } else {
      const newItem: SchemaModel = {
        ...schemaItem,
        idField:
          schemaItem.idField ?? item.fields.find((field) => field.isId)?.name,
        displayFields: schemaItem.displayFields ?? [
          item.fields.find((field) => field.isId)?.name,
        ],
        fields: [],
      };
      item.fields.forEach((field) => {
        const schemaField = schemaItem.fields.find(
          (item) => item.name === field.name,
        );
        if (!schemaField) {
          newItem.fields.push(handleNewField(field, schemaItem.name));
        } else {
          newItem.fields.push({
            ...schemaField,
            ...getOriginalField(field, schemaItem.name),
          });
        }
      });
      newItem.fields.sort((a, b) => a.order - b.order);
      newSchema.models.push(newItem);
    }
  });
  return newSchema;
}

function checkIdFieldExist(model: Model) {
  return !!model.fields.find((field) => field.isId);
}

function handleNewModel(model: Model) {
  const newItem: SchemaModel = {
    id: model.name,
    name: getTitle(model.name),
    idField: model.fields.find((field) => field.isId)!.name,
    displayFields: [model.fields.find((field) => field.isId)!.name],
    create: true,
    update: true,
    delete: true,
    fields: [],
  };
  model.fields.forEach((field) => {
    newItem.fields.push(handleNewField(field, model.name));
  });
  return newItem;
}

const defaultField = ['id', 'createdAt', 'updatedAt'];

function handleNewField(field: Field, modelName: string): SchemaField {
  return {
    ...getOriginalField(field, modelName),
    title: getTitle(field.name),
    create: !defaultField.includes(field.name) && !field.relationField,
    update: !defaultField.includes(field.name) && !field.relationField,
    read: true,
    filter: true,
    sort: true,
    order: -1,
  };
}

function getTitle(id: string) {
  const split = id.split(/(?=[A-Z])/);
  split[0] = split[0].charAt(0).toUpperCase() + split[0].slice(1);
  return split.join(' ');
}

function getOriginalField(
  field: Field,
  modelName: string,
): Omit<Field, 'relation'> & { id: string } {
  delete field.relation;
  return {
    id: modelName + '.' + field.name,
    ...field,
  };
}
