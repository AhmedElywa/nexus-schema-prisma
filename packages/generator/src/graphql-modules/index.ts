import { Options } from '@paljs/types';
import { writeFileSync, readFileSync, existsSync } from 'fs';
import { DMMF } from '../schema';
import { createQueriesAndMutations } from './CreateQueriesAndMutations';
import { Generators } from '../Generators';

export class GenerateModules extends Generators {
  constructor(schemaPath: string, customOptions?: Partial<Options>) {
    super(schemaPath, { output: 'src/app', ...customOptions });
  }

  async run() {
    await this.createModules();
    this.createApp();
  }

  private indexPath = this.output('app.module.ts');
  private index = existsSync(this.indexPath)
    ? readFileSync(this.indexPath, { encoding: 'utf-8' })
    : defaultAppContent;
  private appModules: string[] = getApplication(this.index);

  private async createModules() {
    const models = await this.models();
    const datamodel = await this.datamodel();
    models.forEach((model) => {
      let modules: string[] = ['CommonModule'];
      let extendsTypes = '';

      if (!this.appModules.includes(model.name + 'Module')) {
        this.appModules.push(model.name + 'Module');
        this.index = `import { ${model.name}Module } from './${model.name}/${model.name}.module';${this.index}`;
      }

      let fileContent = `type ${model.name} {`;

      const dataModel = datamodel.models.find(
        (item) => item.name === model.name,
      );
      model.fields.forEach((field) => {
        if (!this.excludeFields(model.name).includes(field.name)) {
          const dataField = dataModel?.fields.find(
            (item) => item.name === field.name,
          );
          if (
            dataField?.kind === 'object' &&
            dataField.relationFromFields.length > 0
          ) {
            if (!modules.includes(dataField.type + 'Module') && model.name !== dataField.type) {
              modules.push(dataField.type + 'Module');
              extendsTypes += `extend type ${dataField.type} {`;
              models
                .find((item) => item.name === dataField.type)
                ?.fields.filter((item) => item.outputType.type === model.name)
                .forEach((item) => {
                  extendsTypes = getField(item, extendsTypes);
                });

              extendsTypes += `}\n\n`;
            }

            fileContent = getField(field, fileContent);
          } else if (
            dataField?.kind !== 'object' ||
            modules.includes(dataField.type + 'Module') ||
            model.name === dataField.type
          ) {
            fileContent = getField(field, fileContent);
          }
        }
      });
      fileContent += `}\n\n`;

      fileContent += extendsTypes;
      this.createFiles(model.name, fileContent);
    });
  }

  getOperations(model: string) {
    const exclude = this.excludedOperations(model);
    return createQueriesAndMutations(
      model,
      this.smallModel(model),
      exclude,
      this.options.onDelete,
    );
  }

  private createFiles(
    model: string,
    content: string,
  ) {
    const operations = this.getOperations(model);

    this.mkdir(this.output(model));

    let resolvers = '';
    let middlewares = '';

    if (!this.disableQueries(model)) {
      resolvers += operations.queries.resolver;
      content += operations.queries.type;
      middlewares += `Query: { "*": [addSelect] },`;
    }
    if (!this.disableMutations(model)) {
      resolvers += operations.mutations.resolver;
      content += operations.mutations.type;
      middlewares += `Mutation:{ "*": [addSelect] },`;
    }

    this.createResolver(resolvers, model);

    this.createTypes(content, model);

    writeFileSync(
      this.output(model, `${model}.module.ts`),
      this.formation(
        getModule(model, middlewares),
        'babel-ts',
      ),
    );
  }

  private createTypes(content: string, model: string) {
    content = `import { gql } from 'graphql-modules';

      export default gql\`
      ${content}
      \`;
      `;

    writeFileSync(
      this.output(model, 'typeDefs.ts'),
      this.formation(content, 'babel-ts'),
    );
  }

  private createResolver(resolvers: string, model: string) {
    if (resolvers) {
      resolvers = `export default {
        ${resolvers}
      }
        `;
      writeFileSync(
        this.output(model, 'resolvers.ts'),
        this.formation(resolvers, 'babel-ts'),
      );
    }
  }

  private createApp() {
    writeFileSync(
      this.indexPath,
      this.formation(App(this.appModules, this.index), 'babel-ts'),
    );
  }
}

const getModule = (
  name: string,
  middlewares: string,
) => {
  return `import { createModule } from 'graphql-modules';
import typeDefs from './typeDefs';
import resolvers from './resolvers';
import { addSelect } from '../common/addSelect';

export const ${name}Module = createModule({
  id: '${name}',
  typeDefs,
  resolvers,
  ${
    middlewares
      ? `middlewares: {
    ${middlewares}
  },
  `
      : ''
  }
});
`;
};

const getField = (field: DMMF.SchemaField, content: string) => {
  content += `
    ${field.name}`;
  if (field.args.length > 0) {
    content += '(';
    field.args.forEach((arg) => {
      content += `${arg.name}: ${arg.inputTypes[0].type}
              `;
    });
    content += ')';
  }
  content += `: ${
    field.outputType.isList
      ? `[${field.outputType.type}!]!`
      : field.outputType.type + (field.isRequired ? '!' : '')
  }`;
  return content;
};

const App = (modules: string[], index: string) => {
  const importObject = index.match(/modules:[\S\s]*?]/);
  if (importObject) {
    const modulesMatch = importObject[0].match(/\[[\S\s]*?]/);
    if (modulesMatch) {
      return index.replace(
        modulesMatch[0],
        JSON.stringify(modules).replace(/"/g, ''),
      );
    }
  }
  return '';
};

const getApplication = (text: string) => {
  const importObject = text.match(/modules:[\S\s]*?]/);
  if (importObject) {
    const modules = importObject[0].match(/\[([\S\s]*?)]/);
    if (modules) {
      return modules[1]
        .split(',')
        .filter((a) => a)
        .map((a) => a.replace(/\s/g, ''));
    }
  }
  return ['CommonModule'];
};

const defaultAppContent = `import { createApplication } from 'graphql-modules';
import { CommonModule } from './common/common.module';

export const application = createApplication({
  modules: [CommonModule],
});`;
