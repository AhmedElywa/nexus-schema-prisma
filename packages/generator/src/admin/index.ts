import { writeFileSync } from 'fs';
import { format } from 'prettier';
import { GeneratorOptions, AdminPagesOptions, SchemaObject } from '@paljs/types';
import { createGraphql } from './createGraphql';
import { mergeSchema } from './mergeSchema';
import { ConvertSchemaToObject } from '@paljs/schema';
import { createFile } from './createFile';

const defaultOptions: GeneratorOptions = {
  prismaName: 'prisma',
  output: './src/graphql',
  excludeFields: [],
  excludeModels: [],
  excludeFieldsByModel: {},
  excludeQueriesAndMutations: [],
  excludeQueriesAndMutationsByModel: {},
};

export class UIGenerator {
  schema: SchemaObject;

  constructor(schemaPath: string | string[]) {
    this.schema =
      typeof schemaPath === 'string'
        ? new ConvertSchemaToObject(schemaPath).run()
        : this.mergeSchemas(schemaPath.map((path) => new ConvertSchemaToObject(path).run()));
  }

  mergeSchemas(schemas: SchemaObject[]) {
    const schema: SchemaObject = { models: [], enums: [] };
    schemas.forEach((s) => {
      schema.models.push(...s.models);
      schema.enums.push(...s.enums);
    });
    return schema;
  }

  buildSettingsSchema(path = 'adminSettings.json', backAsText?: boolean) {
    const newSchema = mergeSchema(this.schema, path);
    const fileContent = format(`${JSON.stringify(newSchema)}`, {
      singleQuote: true,
      semi: false,
      trailingComma: 'all',
      parser: 'json',
    });

    !backAsText && writeFileSync(path, fileContent);

    return newSchema;
  }

  generateGraphql(customOptions?: Omit<Partial<GeneratorOptions>, 'nexusSchema'>) {
    const options: GeneratorOptions = {
      ...defaultOptions,
      ...customOptions,
    };
    return createGraphql(this.schema, options);
  }

  generateAdminPages(options?: AdminPagesOptions) {
    const content = options?.pageContent ?? page;
    const generatedText: Record<string, string> = {};
    this.schema.models
      .filter((model) => !options?.models || options?.models.includes(model.name))
      .forEach((model) => {
        const fileContent = format(content.replace(/#{id}/g, model.name), {
          semi: true,
          trailingComma: 'all',
          singleQuote: true,
          printWidth: 120,
          tabWidth: 2,
          parser: 'babel-ts',
        });
        if (options?.backAsText) {
          generatedText[model.name] = fileContent;
        } else {
          createFile(options?.outPut ?? 'src/pages/admin/models/', `${model.name}.tsx`, fileContent);
        }
      });
    return generatedText;
  }
}

const page = `
import React from 'react';
import PrismaTable from 'components/PrismaTable';

const #{id}: React.FC = () => {
  return <PrismaTable model="#{id}" />;
};

export default #{id};
`;
