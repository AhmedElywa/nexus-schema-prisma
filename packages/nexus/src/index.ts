import { PrismaSelect } from '@paljs/plugins';
import type { DMMF } from '@paljs/types';
import { objectType, plugin } from 'nexus';
import { NexusAcceptedTypeDef } from 'nexus/dist/builder';
import { adminNexusSchemaSettings } from './admin';
import type { Settings } from './settings';
import { getScalars } from './defaultScalar';
import { AllNexusOutputTypeDefs } from 'nexus/dist/definitions/wrapping';

export { Settings };

export const paljs = (settings?: Settings) =>
  plugin({
    name: 'paljs',
    description:
      'paljs plugin to add Prisma select to your resolver and prisma admin queries and mutations and all models input types',
    onInstall(builder) {
      const nexusSchemaInputs: (NexusAcceptedTypeDef | AllNexusOutputTypeDefs)[] = [
        objectType({
          name: 'BatchPayload',
          definition(t) {
            t.nonNull.int('count');
          },
        }),
        ...getScalars(settings?.excludeScalar),
      ];

      if (settings?.includeAdmin) {
        nexusSchemaInputs.push(...adminNexusSchemaSettings(settings?.adminSchemaPath));
      }
      for (const type of nexusSchemaInputs) {
        builder.addType(type);
      }
    },
    onCreateFieldResolver() {
      return async (root, args, ctx, info: any, next) => {
        ctx.select = new PrismaSelect(info, {
          ...settings?.prismaSelectOptions,
          dmmf: settings?.prismaSelectOptions?.dmmf as Omit<DMMF.Document, 'schema'>[],
        });
        return await next(root, args, ctx, info);
      };
    },
  });
