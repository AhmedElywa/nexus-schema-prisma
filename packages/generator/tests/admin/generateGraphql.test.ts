import { join } from 'path';
import { UIGenerator } from '../../src';

describe('Generate Graphql queries And mutations for Frontend', () => {
  const schemaPath = join(__dirname, '../schemas/schema.prisma');
  it('Should back with all models', () => {
    const generator = new UIGenerator(schemaPath);
    expect(generator.generateGraphql({ backAsText: true })).toMatchSnapshot();
  });

  it('Should back User model graphql with custom exclude', () => {
    const generator = new UIGenerator(schemaPath);
    expect(
      generator.generateGraphql({
        models: ['User'],
        backAsText: true,
        excludeFields: ['password'],
        excludeQueriesAndMutations: ['findCount', 'createOne'],
        excludeQueriesAndMutationsByModel: { User: ['updateMany'] },
      }),
    ).toMatchSnapshot();
  });
});
