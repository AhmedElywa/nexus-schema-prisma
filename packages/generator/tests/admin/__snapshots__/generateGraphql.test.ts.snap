// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Generate Graphql queries And mutations for Frontend Should back User model graphql with custom exclude 1`] = `{}`;

exports[`Generate Graphql queries And mutations for Frontend Should back with all models 1`] = `{}`;
